x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteForecastAction.java

```diff
@@ -0,0 +1,111 @@
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.DeleteByQueryAction;
+import org.elasticsearch.index.reindex.DeleteByQueryRequest;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.ml.action.DeleteForecastAction;
+import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.core.ml.job.messages.Messages;
+import org.elasticsearch.xpack.core.ml.job.persistence.AnomalyDetectorsIndex;
+import org.elasticsearch.xpack.core.ml.job.results.Forecast;
+import org.elasticsearch.xpack.core.ml.job.results.ForecastRequestStats;
+import org.elasticsearch.xpack.core.ml.job.results.ForecastRequestStats.ForecastRequestStatus;
+import org.elasticsearch.xpack.core.ml.job.results.Result;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.job.persistence.JobResultsProvider;
+
+import java.util.EnumSet;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+
+public class TransportDeleteForecastAction extends HandledTransportAction<DeleteForecastAction.Request,
+    AcknowledgedResponse> {
+
+    private static final String RESULTS_INDEX_PATTERN = AnomalyDetectorsIndex.jobResultsIndexPrefix() + "*";
+    private final Client client;
+    private final JobResultsProvider jobResultsProvider;
+    private static final Set<ForecastRequestStatus> DELETABLE_STATUSES =
+        EnumSet.of(ForecastRequestStatus.FINISHED, ForecastRequestStatus.FAILED);
+
+    @Inject
+    public TransportDeleteForecastAction(Settings settings, TransportService transportService, ActionFilters actionFilters, Client client,
+                                         JobResultsProvider jobResultsProvider) {
+        super(settings, DeleteForecastAction.NAME, transportService, actionFilters,
+            DeleteForecastAction.Request::new);
+        this.client = client;
+        this.jobResultsProvider = jobResultsProvider;
+    }
+
+    @Override
+    protected void doExecute(Task task, DeleteForecastAction.Request request, ActionListener<AcknowledgedResponse> listener) {
+        final String forecastId = request.getForecastId();
+        final String jobId = request.getJobId();
+        jobResultsProvider.getForecastRequestStats(jobId, forecastId, (forecastRequestStats) -> {
+            if (forecastRequestStats == null) {
+                listener.onFailure(
+                    new ResourceNotFoundException(Messages.getMessage(Messages.REST_NO_SUCH_FORECAST, forecastId, jobId)));
+                return;
+            }
+
+            if (DELETABLE_STATUSES.contains(forecastRequestStats.getStatus())) {
```

Comment:
We could reduce the indentation here by checking if the status is `not` deletable first, then proceed in doing the deletion. I leave it on your preference whether you want to change it.